Improve navigate for service worker hooks

jungkees opened the pull request:
Improve navigate for service worker hooks

Before this patch, there has been no concept that holds reserved
client's information referenced from a service worker during a
navigation (more precisely before the actual global object and its
relevant environment settings object for the resulting resource is
created).

This patch defines an "environment" concept (a supertype object of an
environment settings object) that has an id, a creation URL, a target
browsing context, and an active service worker such that service
workers' FetchEvent hanlders can reference this client information
before the corresponding environmet setings object is actually
created.

The changes include:
 - Create an environment (a reserved client) during a navigation; for
 worker creation requests, create an environment settings object and
 set the fields correspondingly
 - Set request's reserved client to the created environment or
 environment settings object such that fetch (and Handle Fetch) can
 use it
 - Set request's target client id such that fetch (and Handle Fetch)
 can use it
 
 * Note: The reserved client's active service worker is set, if
 available, during the fetch (in service worker's Handle Fetch
 algorithm).

Related issue: https://github.com/w3c/ServiceWorker/issues/870
Related change: https://github.com/whatwg/fetch/pull/383

jungkees commented:
@annevk, I put together the changes based on the outline here:
 - Call flow: https://github.com/w3c/ServiceWorker/issues/870#issuecomment-241341776
 - Expected behavior: https://github.com/w3c/ServiceWorker/issues/870#issuecomment-245559217

I'd appreciate your review.
/cc @domenic

annevk commented:
Note for others interested in this PR, we're discussing the general design in https://github.com/whatwg/fetch/pull/383.

jungkees commented:
@annevk, I basically regarded the environment concept and the environment settings object concept are covariant. So, I defined it as "an environment settings object is an environment." and later in the service worker, I'll use the environment to enumerate all the clients there. Please take a look.

jungkees commented:
Self-note: destroy the environment object after when its values are set up to the resulting environment settings object. (missed it.)

annevk commented on source:
```diff
@@ -77994,12 +78001,30 @@ console.assert(iframeWindow.frameElement === null);
+  optional opaque string <var>reservedId</var>, it must run the following steps:</p>
+
+  <ol>
+   <li><p>If <var>reservedId</var> is present, set <var>settingsObject</var>'s <span>id</span> to
```
`<span>id</span>` probably needs `data-x`?

jungkees replied:
Done.

annevk commented on source:
```diff
@@ -78010,96 +78035,53 @@ console.assert(iframeWindow.frameElement === null);
-      <p>Return <var>execution context</var>.</p>
-
-     </dd>
+     <dd><p>Return <var>executionContext</var>.</p></dd>
```
I like this kind of cleanup, but @domenic needs to approve it. Also, it might be better to do that in a separate editorial-changes-only commit.

jungkees replied:
I'll follow what you and @domenic guide here.

domenic replied:
Yes, let's please remove the reflowing of paragraphs/deleting of blank lines/renaming of variables from this commit.

annevk commented on source:
```diff
@@ -82141,6 +82141,16 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+
+   <li><p>Set <var>request</var>'s <span data-x="concept-request-reserved-client-id">reserved
+   client id</span> to <var>reservedEnvironment</var>'s <span
+   data-x="concept-environment-id">id</span>.</p></li>
```
I don't really understand this. If we create the environment here, doesn't it need to be passed along with the request? Or is the idea that we later find it again through its id?

jungkees replied:
That'd be an option, but I think fetch already has the other states on the request already: the url and the target browsing context.

What service worker needs during dispatching the fetch event is the reserved environment's id to set `FetchEvent.reservedClientId`. And yes, the idea is to let the fetch event handler get the client using e.g. `clients.get(e.reservedClientId)`.

In the similar reason, I proposed adding a target client id internal slot to request, too: https://github.com/whatwg/fetch/pull/383/commits/4511e2e354759261f3cf4d07ee4ecb67ac3110fa As discussed during the f2f, we wanted to capture the target client's id here instead of crossing the process border later when the service worker needs to access the target client in the fetch event dispatch steps.

annevk commented on source:
```diff
@@ -82253,15 +82263,14 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
    <li><p>Run <span>process a navigate response</span> given <var>request</var>,
-   <var>response</var>, <var>navigationType</var>, the <span>source browsing context</span>, and
-   <var>browsingContext</var>.</p></li>
+   <var>response</var>, <var>navigationType</var>, the <span>source browsing context</span>, <var>browsingContext</var>, and <var>reservedEnvironment</var>.</p></li>
```
The wrapping here seems wrong.

jungkees replied:
Fixed.

jungkees commented:
> I'm also a little unsure about naming. We have "environment settings object" which effectively means global object. This adds "environment" which means something like "future/reserved global object".

Open to renaming it. I absolutely want a better name for it if we can get one.

> It's also a little weird still that we grab some state from request to create something new and then grab something from the newly created thing and tack it on request. That just doesn't seem quite right.

I don't think I can help here. The only state I copy from request to environment is the request's url which the newly created environment has to have. (Note that I tried to factor the creation URL out to environment as both environment and environment settings object need this info.) And the only state I copy from the created environment to request is its id, which is the whole idea this patch suggests. I think it makes sense as a navigation works in this way.

annevk commented:
Are you saying that this is also how Chrome would engineer this setup? @wanderview would we use this setup in Firefox?

jungkees commented:
I'll dig into it again, but this is what I found sometime ago: https://github.com/whatwg/html/pull/322#issuecomment-200229641

Basically, Chromium will have to implement something similar to the reserved id, store it somewhere like in the service worker provider and later relate the id to the actual frame id when the frame and the document are created. So, I think, in order to expose the `FetchEvent.reservedClientId` and `FetchEvent.targetClientId`, we'll have to implement the similar logic. It'll be great if @wanderview and @mattto cross-checks if this setup is good to implement. Please check out this simplified flow to help your understanding: https://github.com/w3c/ServiceWorker/issues/870#issuecomment-241341776

Also, I'd appreciate your comments on calling the match service worker registration algorithm twice: once for main resource fetch and the other for attaching a controller after the global is created. This is due to the fact that it's not easy to pass the captured state around between navigate/fetch/SW.

jungkees commented:
Having thought a bit more on this, matching the registration twice is awkward. I think we can do better as we now have the environment concept where we can store the matched registration. For this, we'd have to pass this environment along with the request as @annevk asked above.

So, the idea is:
 - Match a service worker registration with request's url
 - Create an environment and set all their internal slots properly including the service worker registration internal slot to the match registration
 - Check (in Handle Fetch) if the reserved client's service worker registration is not null and its active worker is not null to dispatch fetch event. (For subresource fetch, client's registration should be used instead of the reserved client's registration.)

For worker clients, we can basically do the same and will pass the entire environment settings object (instead of an environment) that is set with the matched registration to fetch. This means the type of "reserved client" internal slot to request should be "null or an environment or an environment settings object".

I think this matches the implementations more precisely, at least Chromium. @wanderview, would it make sense for Gecko? (Note that this doesn't change the originally proposed behavior in this PR. Both matching the registration twice and matching it once / storing in an environment will use the latest active worker always.)

jungkees commented:
@mattto, @kinu I'd appreciate your comments on this.

jungkees commented:
Having reviewed Chromium's navigation part of the code again, I believe the structure I'm trying to spec in this PR would be implementable.

The current implementation is explained in https://github.com/whatwg/html/pull/322#issuecomment-200229641 The gist is when it initializes the request handler for a navigation request (`ServiceWorkerRequestHandler::InitializeForNavigation` - largely corresponding to initializing a request to pass to Handle Fetch) it already sets up the |provider_host|'s client (request's client in the spec) to the new reserved client - which the spec will do with the change. The current implementation just ignores this value by setting the `FetchEvent's clientId` to null if it's a navigation request.

So for navigation requests, Chromium will need to change to set the request handler's provider_host to the source client instead of the reserved client. And add a reserved_provider_host or similar state to grab the pre-created (i.e. reserved) client and the target_provider_host's client_id.

Still would like to check what Gecko does and whether the proposed structure is implementable there.

domenic commented on source:
```diff
@@ -77994,12 +78001,31 @@ console.assert(iframeWindow.frameElement === null);
 
   <h4>Script settings for browsing contexts</h4>
 
+  <p>When the user agent is required to <dfn>set up an environment</dfn>, given an
```
Since this is browsing context specific, then it should probably have that in the name, like "set up a browsing context environment settings object" does. So "set up a browsing context environment".

jungkees replied:
Indeed this algorithm isn't browsing context specific. It's invoked from both set up a browsing context environment settings object and set up a worker environment settings object. I think this algorithm definition needs to move to a better place?

domenic commented on source:
```diff
@@ -77994,12 +78001,31 @@ console.assert(iframeWindow.frameElement === null);
+  optional opaque string <var>reservedId</var>, it must run the following steps:</p>
+
+  <ol>
+   <li><p>If <var>reservedId</var> is present, set <var>settingsObject</var>'s <span
```
s/is present/is given/

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -78487,6 +78483,16 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
  http://lxr.mozilla.org/mozilla/source/dom/public/idl/base/nsIDOMWindowInternal.idl - DOM level 0
 -->
 
+  <p>A <code>Window</code> object has an <dfn data-export=""
```
Why move this? I think it should stay where it is (in the w-nodev div) and the active worker should also go there.

jungkees replied:
I thought it would've been more consistent to other part of the spec such as WorkerGlobalScope definition where the internal slots are defined before the method/attribute description section. But I found it's not very consistent across the document. Reverted this line. (Note that active worker isn't defined in the global object any long, so it's removed here.)

domenic commented on source:
```diff
@@ -78487,6 +78483,16 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+
+  <p>A <code>Window</code> object has an associated <dfn data-export="" data-dfn-for="Window"
+  data-x="concept-Window-active-worker">active worker</dfn> (null or a <span
+  data-x="dfn-service-worker">service worker</span>). It is initially null.</p>
```
We should call this "active service worker"; it is quite confusing to me otherwise.

jungkees replied:
This state has been moved to the environment concept. So, it's called environment's service worker now. Not sure whether "active service worker" would be a better name though. This internal slot basically holds the value of either null or a service worker in "active" state only.

domenic commented on source:
```diff
@@ -82456,6 +82472,32 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+      <ol>
+       <li><p>Let <var>registration</var> be the result of running <span
+       data-x="scope-match-algorithm">match service worker registration</span> with the
+       <code>Document</code>’s <span>URL</span>.</p></li>
```
No curly quotes

jungkees replied:
Addressed. Thanks for spotting.

jungkees commented:
Thanks @domenic. I'll address your comments and this change (https://github.com/whatwg/html/pull/1776#issuecomment-251319419) after checking with @wanderview if Gecko has any concerns.

jungkees commented:
@wanderview, I'd like to check with you this spec change is feasible with Gecko. I'd appreciate your comments on the below questions.

Gecko's using a document id to identify a client. It generates a new document id in `nsDocShell::ChannelIntercepted` and passes it to the fetch event dispatch procedure where it seems to be planing to use this document id to set `FetcheEvent.reservedClientId` according to the comments in the code. I'd like to ask and discuss around the possibilities below:

 - It seems the current client getter methods (e.g. `clients.get(id)`) are not able to grab a reserved client in fetch event's handlers as only the already-being-created documents (i.e. those set with `AddObserver(.."service-worker-get-client"..)` ) are observable in the client list. So, this needs changes I guess.
 - Can Gecko define and use a new object that has the id, creation URL, reference to target browsing context and matched service worker registration instead of using a document id?
   (As an aside, I think this environment concept associated with a global is a better fit to represent what's controlled than a document as we have both window client and worker client concepts.)
 - Can we do the registration matching at the beginning of a navigation (before calling fetch) as proposed in https://github.com/whatwg/html/pull/1776#issuecomment-251319419?
 - Gecko will need to capture the target DocShell's associated Window's client id during a navigation. (Chromium will need to add this as well.)

jungkees commented:
Had a chat with @wanderview. We agreed to proceed with this PR. I'll address the comments given so far and the changes outlined in https://github.com/whatwg/html/pull/1776#issuecomment-250465276, and ask review.

jungkees commented:
@annevk @domenic, I uploaded new snapshots having addressed the comments and the additional changes discussed. I believe this outlined flow will help your understanding: https://github.com/w3c/ServiceWorker/issues/870#issuecomment-241341776

* Summary
 - An environment has |id|, |creation URL|, |target browsing context|, |service worker|.
 - An environment is created during a navigation to represent a reserved client to SW.
 - Later in the navigation step when the environment settings object is create, the reserved environment's states are copied to the environment settings object. (that environment is destroyed here.)
 - For other occasions where an environment settings object is created (e.g. initial browsing context creation, etc.), the environment part of the states are set to initial values. (|id| is set to a new opaque string.)
 - Worker creation is similar, but in this case, an environment settings object is created before invoking fetch, so no separate environment is created.
 - Service worker registration matching is executed in SW's Handle Fetch algorithm (WIP).
 - Once a active service worker is set to the reserved client(environment), that service worker's registration's waiting worker won't promote until all the client including this client will be unloaded; this resolves my previous concern about having different service worker versions for the main resource request and its subresource requests.
 - SW will expose the reserved client's id, target client's id as well as the (source) client's id to fetch event handlers (WIP).
 - SW's clients-getter-methods will enumerate service worker clients including environments and environment settings objects depending on the given options to the methods.

domenic commented:
I should have time to review this tomorrow, but if you have a chance to remove the re-wrapping and blank-line-removing, that would make things easier to review. It's much better now than it was though :)

jungkees commented:
Sounds great. I updated the snapshot having removed the re-wrapping and blank-line-removing. Please take a look.

domenic commented on source:
```diff
@@ -77983,9 +77990,33 @@ console.assert(iframeWindow.frameElement === null);
 
   <h4>Script settings for browsing contexts</h4>
 
+  <p>When the user agent is required to <dfn>set up an environment</dfn>, given an <span>environment
```
As discussed, this should move. I think a good place is right after the definition of environment (in 8.1.3.1). Also, the argument should be "an environment _environment_", not "an environment settings object _environmentSettings_".

domenic replied:
Additionally, it might be good to split this into two algorithms: "set up a simple environment" which takes an environment and a URL, and "set up an environment from another environment" which takes an environment. Then no call sites would have to pass null explicitly, and there would not be an optional ID parameter.

jungkees replied:
I removed "set up an environment" algorithm as you commented below. Indeed, "set up an environment from another environment" would have had only one call site, so it seems nicer to explicitly set those fields without defining an algorithm for it.

domenic commented on source:
```diff
@@ -86262,8 +86336,27 @@ interface <dfn>NavigatorOnLine</dfn> {
 
-  <p>An <dfn data-export="">environment settings object</dfn> specifies algorithms for obtaining the
-  following:</p>
+  <p>An <dfn data-export="">environment</dfn> is an object that identifies the settings of the
```
I think this would end up looking a lot clearer as a `<dl>` like environment settings object. The only difference is that instead of "specifies algorithms for obtaining the following," you'd say something like "has the following fields"

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -86262,8 +86336,27 @@ interface <dfn>NavigatorOnLine</dfn> {
+  string), which uniquely identifies itself. An <span>environment</span> has a <dfn
+  data-x="concept-environment-creation-url" data-export="" data-dfn-for="environment">creation
+  URL</dfn> (an <span>absolute URL</span>) representing the location of the resource with which the
+  <span>environment</span> is associated. Note that this URL might be distinct from the
```
Notes should go in `<p class="note">` (and not include the words "Note that" inside such paragraphs). I guess that was note the case before but it would be nice to fix while you're here :)

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -86262,8 +86336,27 @@ interface <dfn>NavigatorOnLine</dfn> {
+  data-x="dfn-control">controls</span> it. It is initially null.</p>
+
+  <p>An <dfn data-export="">environment settings object</dfn> is an <span>environment</span> that
+  specifies algorithms for obtaining the following:</p>
```
Maybe "additionally specifies algorithms for"

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -86262,8 +86336,27 @@ interface <dfn>NavigatorOnLine</dfn> {
+  <span>environment</span> has a <dfn data-x="concept-environment-active-service-worker"
+  data-export="" data-dfn-for="environment">active service worker</dfn> (null or a <span
+  data-x="dfn-service-worker">service worker</span>) which <span
+  data-x="dfn-control">controls</span> it. It is initially null.</p>
```
Instead of stating that some things are initially null, I would add a sentence saying when these fields are set. E.g. "These fields are all set by the set up an environment algorithm" or "These fields are all set when the environment is created". Or just leave it out I guess, since that's kind of obvious :)

jungkees replied:
These fields are set during the setup of an environment or an environment settings object. So, I just left it out as you commented.

domenic commented on source:
```diff
@@ -82129,6 +82163,21 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
    and unset <var>request</var>'s <span>omit-<code>Origin</code>-header flag</span>.</p></li>
 
    <li>
+    <p>Create a new <span>environment</span> <var>reservedEnvironment</var>, and set its <span
```
Hmm now I am questioning whether you need the "set up an environment" algorithm(s) at all. You can just create them and set the fields like you do here. That seems pretty nice too.

jungkees replied:
Addressed this comment: set the fields in the relevant steps instead of defining/calling an algorithm.

jungkees commented:
@domenic, I updated the snapshot having addressed your comments. Please take a look.

domenic commented on source:
```diff
@@ -82129,6 +82157,21 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
 
    <li>
+    <p>Create a new <span>environment</span> <var>reservedEnvironment</var>, and set its <span
+    data-x="concept-environment-id">id</span> to a new unique opaque string, its <span>creation
```
creation URL is not linking correctly; needs a data-x

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -82129,6 +82157,21 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+    data-x="concept-environment-target-browsing-context">target browsing context</span> to
+    <var>browsingContext</var>.</p>
+
+    <p class="note">The created environment's active service worker is set in the <span
```
Link active service worker?

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -82129,6 +82157,21 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+
+    <p class="note">The created environment's active service worker is set in the <span
+    data-x="on-fetch-request-algorithm">handle fetch</span> algorithm during the subsequent fetch
+    flow if its creation URL matches a service worker registration. <ref spec="SW"></p>
```
Link creation URL?

domenic commented on source:
```diff
@@ -82443,6 +82490,27 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+       <ol>
+        <li><p>If <var>browsingContext</var> has an <span>opener browsing context</span> and
+        <var>request</var>'s <span>client</span> is a <span>non-secure context</span>, then
+        <span>disown</span> <var>browsingContext</var>'s opener.</p></li>
```
Need a data-x to link the disown. Also it looks like the span is not closed.

domenic commented on source:
```diff
@@ -86262,8 +86330,58 @@ interface <dfn>NavigatorOnLine</dfn> {
 
-  <p>An <dfn data-export="">environment settings object</dfn> specifies algorithms for obtaining the
-  following:</p>
+  <p>An <dfn data-export="">environment</dfn> is an object that identifies the settings of the
```
I think the phrasing "the settings of **a** current or potential execution environment" is better.

jungkees replied:
Done. Thanks.

domenic commented on source:
```diff
@@ -86262,8 +86330,58 @@ interface <dfn>NavigatorOnLine</dfn> {
+
+   <dd>
+
+    <p>An opaque string that uniquely identifies itself.</p>
```
s/itself/the environment/

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -86262,8 +86330,58 @@ interface <dfn>NavigatorOnLine</dfn> {
+    <p>An <span>absolute URL</span> that represents the location of the resource with which the
+    <span>environment</span> is associated.</p>
+
+    <p class="note">This URL might be distinct from the <span>environment settings object</span>'s
```
Hmm can we prefix this with "**In the case of an environment settings object,** this URL might be distinct..."

jungkees replied:
Yes, it's more accurate with the phrase. Done.

domenic commented on source:
```diff
@@ -86262,8 +86330,58 @@ interface <dfn>NavigatorOnLine</dfn> {
+
+   <dd>
+
+    <p>null or a target <span>browsing context</span> for a <span data-x="navigation-request">navigation request</span>.</p>
```
Over 100 chars

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -86262,8 +86330,58 @@ interface <dfn>NavigatorOnLine</dfn> {
+   <dd>
+
+    <p>null or a <span data-x="dfn-service-worker">service worker</span> that <span
+    data-x="dfn-control">controls</span> it.</p>
```
s/it/the environment

jungkees replied:
Done.

domenic commented on source:
```diff
@@ -3883,6 +3889,7 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
 
     <ul class="brief">
      <li><dfn data-x-href="https://w3c.github.io/webappsec-secure-contexts/#settings-object">Is environment settings object a secure context?</dfn></li>
+     <li><dfn data-x="secure-context" data-x-href="https://w3c.github.io/webappsec-secure-contexts/#secure-context">secure context</dfn></li>
```
FYI for terms like "secure context" and "service worker" that are not very ambiguous, there's no need to have data-x. You can just do `<dfn data-x-href="...">secure context</dfn>` and then at the reference site do `<span>service worker</span>`. But this way is fine too, no need to change.

jungkees replied:
I just left it out but wonder if there's any guideline that one can determine whether to use data-x for a definition.

annevk replied:
The guideline is basically whether the term is unique enough to stand out on its own. I think here it would be fine to leave out data-x="".

jungkees commented:
Thanks for the comments, @domenic. I uploaded a new snapshot having addressed them. /cc @annevk

jungkees commented on source:
```diff
@@ -78105,6 +78105,34 @@ console.assert(iframeWindow.frameElement === null);
+     service worker</span> to <var>reserved environment</var>'s <span
+     data-x="concept-environment-active-service-worker">active service worker</span>.</p></li>
+
+     <li><p>Destroy <var>reserved environment</var>.</p></li>
```
I was worrying about the situation where we get two objects (both the environment and the environment settings object) from a single id, which can happen if the environment is not destroyed properly right after this step. Or should I just reset the id of the environment to the empty string here?

annevk replied:
I see. Setting the id and adding a note about how it's observable would be good. "Destroy" doesn't really have any meaning, so I'd rather not use it.

annevk commented on source:
```diff
@@ -78105,6 +78105,34 @@ console.assert(iframeWindow.frameElement === null);
+   </li>
+
+   <li><p>Otherwise, set <var>settings object</var>'s <span
+   data-x="concept-environment-id">id</span> to a new unique opaque string, <var>settings
```
I don't think so. What might be problematic is if folks start assuming the string is UTF-8-safe or even ASCII-safe, depending on what user agents do. Is it an option to return something other than a string that can only be used to check for object identity? E.g., a symbol?

jungkees replied:
> only be used to check for object identity

FWIW, I think this is the only purpose of the *id* itself. I don't think I'm familiar with what you meant by a symbol, though. Could you explain more? Would it be shown as just a string to JS?

annevk replied:
As a Symbol: https://tc39.github.io/ecma262/#sec-symbol-objects. They're somewhat more opaque than strings so might be more suitable here. Not a 100% sure though and it also depends a bit on what happens with the ID.

domenic replied:
Symbol doesn't work great across realms, without basically giving it a string name (like `"Symbol.iterator"`) in which case we're back to strings. String is probably the best we'll be able to do.

annevk replied:
If these IDs are meant to work across realms, that makes sense.

annevk commented on source:
```diff
@@ -3866,10 +3868,14 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
 
     <ul class="brief">
+     <li><dfn data-x="dfn-active-worker" data-x-href="https://w3c.github.io/ServiceWorker/#dfn-active-worker">active worker</dfn></li>
      <li><dfn data-x="dfn-client-message-queue" data-x-href="https://slightlyoff.github.io/ServiceWorker/spec/service_worker/#dfn-client-message-queue">client message queue</dfn></li>
```
If we land this PR first before the other PR I wrote where I update links we should also update this link here.

jungkees replied:
Addressed.

annevk commented on source:
```diff
@@ -86262,8 +86340,60 @@ interface <dfn>NavigatorOnLine</dfn> {
+  current or potential execution environment. An <span>environment</span> has the following
+  fields:</p>
+
+  <dl>
```
Since this is a new `<dl>` I feel like we should remove a whole bunch of whitespace. E.g., when ` <dd>` only contains a single `<p>` it can be all together. No newline between `<dt>` and `<dd>`, only between `<dd>` and `<dt>`. What do you think?

jungkees replied:
Referencing the contributing.md,

> If a "block" element contains a single "block" element, do not put it on a newline. Do not indent for anything except a new "block" element.

I think that makes sense for new works and changed them as such.

annevk commented on source:
```diff
@@ -86262,8 +86340,60 @@ interface <dfn>NavigatorOnLine</dfn> {
+   <dd>
+
+    <p>null or a target <span>browsing context</span> for a <span
+    data-x="navigation-request">navigation request</span>.</p>
```
This should say "Null" I think. We don't start with a lowercase letter in these `<dd>`s anywhere else either. Same for below.

jungkees replied:
Done.

annevk commented:
LGTM with those last few nits fixed. I assume we don't have a testing plan here because this is is about new stuff? That's probably okay.

jungkees commented:
@annevk, I uploaded a new snapshot having addressed those comments. Please take a look.

annevk commented:
I'm going to merge this now and then work with @jungkees today/tomorrow to get the corresponding change to Fetch landed as soon as possible. Either way there'll be some broken URLs.

annevk commented:
Actually, @jungkees could you write a new commit message perhaps?

jungkees commented:
@annevk, I'm not sure but presume you'll squash all the commits to a single commit, right? I have no idea if I can amend any commit message myself, so I updated the commit message in the OP. Let me know if I can help in another way.

jungkees commented:
@annevk @domenic, thanks for having reviewed this PR.

annevk commented:
Thanks for writing it! 😊
